internal/worker/concurrency_auto_scaler_test.go

```diff
@@ -0,0 +1,392 @@
+// Copyright (c) 2017-2021 Uber Technologies Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package worker
+
+import (
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/stretchr/testify/assert"
+	"github.com/uber-go/tally"
+	"go.uber.org/atomic"
+	"go.uber.org/goleak"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zaptest/observer"
+
+	"go.uber.org/cadence/.gen/go/shared"
+	"go.uber.org/cadence/internal/common"
+)
+
+const (
+	testTickTime   = 1 * time.Second
+	testTimeFormat = time.TimeOnly
+)
+
+func createTestConcurrencyAutoScaler(t *testing.T, logger *zap.Logger, clock clockwork.Clock) *ConcurrencyAutoScaler {
+	return NewConcurrencyAutoScaler(ConcurrencyAutoScalerInput{
+		Concurrency: &ConcurrencyLimit{
+			PollerPermit: NewResizablePermit(100),
+			TaskPermit:   NewResizablePermit(1000),
+		},
+		Cooldown:                 2 * testTickTime,
+		Tick:                     testTickTime,
+		PollerMaxCount:           200,
+		PollerMinCount:           50,
```

Comment:
somewhat odd that `PollerPermit` starts at a different value than `PollerMinCount`, since I don't believe that'll ever be the case in practice?

it does seem harmless though, and kinda simplifies the "scale down" tests... just not sure that's worth breaking the normal pattern to achieve.

Reply:
In practice, we observe the `MaxConcurrentPollers` set in the top level and use that as the initial poller count.